### PR TITLE
Bug 352867 multi-threading hardening Workspace.tree / ElementTree

### PR DESCRIPTION
**Pull request by jukzi**
* unify SaveManager.sortTrees and ElementTreeWriter.sortTrees to have a
single sort implementation,
* added some missing volatile/final/synchronized
* improved errormessage, if it should still occur.
* added junit for sorting

Signed-off-by: Joerg Kubitz <jkubitz-eclipse@gmx.de>

### CODE REVIEW

File: bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/SaveManager.java [hunk 1]
```diff
@@ -1495,27 +1496,29 @@ protected void snapTree(ElementTree tree, IProgressMonitor monitor) throws CoreE
 	}
 
 	/**
-	 * Sorts the given array of trees so that the following rules are true:
```
Comment (iloveeclipse) [inline bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/SaveManager.java:1498]: This comment is outdated now, or why is this removed? It contains information we don't have anywhere else.

File: bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/Workspace.java [hunk 2]
```diff
@@ -2160,8 +2160,12 @@ public Resource newResource(IPath path, int type) {
 	 * modifications to the tree.
 	 */
 	public ElementTree newWorkingTree() {
-		tree = tree.newEmptyDelta();
-		return tree;
+		// synchronized for atomic swap. Should have already synchronized by
+		// getWorkManager().checkIn/checkout, but it's not guaranteed
```
Comment (iloveeclipse) [inline bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/Workspace.java:2164]: According to the comment, we might have **two** locks now involved? This is dangerous.

File: bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/SaveManager.java (hunk 1 above)
Comment (jukzi) [reply bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/SaveManager.java:1498]: 1. information was wrong. As proofed with junit the first tree could have a parent.
2. the valid information extended in the param section

File: bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/Workspace.java (hunk 2 above)
Comment (jukzi) [reply bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/Workspace.java:2164]: newEmptyDelta() does not lock anything beside its individual synchronize.

File: bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/watson/ElementTree.java [hunk 3]
```diff
@@ -79,65 +80,70 @@
...
+		this.tree = newTree;
 	}
 
 	/**
 	 * Creates a new empty delta element tree having the
 	 * given tree as its parent.
 	 */
 	protected ElementTree(ElementTree parent) {
-		if (!parent.isImmutable()) {
-			parent.immutable();
```
Comment (iloveeclipse) [inline bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/watson/ElementTree.java:131]: why is this now removed?

Comment (EclipseNewb) [reply bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/watson/ElementTree.java:131]: that moved to the caller  newEmptyDelta():624 to get the constructor at the first place.

**Comment by EclipseNewb** [comment]
unrelated fail testBuilderDeltaUsingRelaxedRuleBug343256 fails randomly #19

File: bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/watson/ElementTreeWriter.java [hunk 4]
```diff
@@ -78,65 +79,23 @@ public Object readData(IPath path, DataInput input) {
...
-	 * 	 - No tree has an ancestor with a greater index in the array.
-	 * If there are no missing parents in the given trees array, this means
-	 * that in the resulting array, the i'th tree's parent will be tree i-1.
-	 * The input tree array may contain duplicate trees.
-	 * The sort order is written to the given output stream.
+	 * Sorts the given array of trees and The sort order is written to the given
+	 * output stream.
 	 */
 	protected ElementTree[] sortTrees(ElementTree[] trees, DataOutput output) throws IOException {
+		ElementTree[] sorted = SaveManager.sortTrees(trees);
```
Comment (iloveeclipse) [inline bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/watson/ElementTreeWriter.java:86]: `sortTrees()` can return null. The sort that was before here would throw IOException in that case, now NP exception. This is a problem, because the two callers of `org.eclipse.core.internal.watson.ElementTreeWriter.writeDeltaChain(ElementTree[], IPath, int, DataOutput, IElementComparator)` have dedicated catches for IO exception but nothing for NP.

That must be fixed.
